modules/ui/src/main/scala/lucuma/ui/forms/FormInputEV.scala

```diff
@@ -53,92 +55,135 @@ final case class FormInputEV[EV[_], A](
 )(implicit val ev: ExternalValue[EV])
     extends ReactProps[FormInputEV[Any, Any]](FormInputEV.component) {
-  def valGet: String = ev.get(value).foldMap(format.reverseGet)
-  def valSet(s: String): Callback =
-    format.getOption(s).map(ev.set(value)).getOrEmpty
-  val onBlurC: InputEV.ChangeCallback[String]   =
-    (s: String) => format.getOption(s).map(onBlur).getOrEmpty
-  val onChangeC: InputEV.ChangeCallback[String] =
-    (s: String) => format.getOption(s).map(onChange).getOrEmpty
+
+  def valGet: String = ev.get(value).foldMap(validate.reverseGet)
+
+  def valSet: InputEV.ChangeCallback[A] = ev.set(value)
+
+  def onBlurC(onError: NonEmptyList[String] => Callback): InputEV.ChangeCallback[String] =
+    (s: String) =>
+      validate
+        .getValidated(s)
+        .fold(onError, onBlur)
 
   def withMods(mods: TagMod*): FormInputEV[EV, A] = copy(modifiers = modifiers ++ mods)
 }
 
 object FormInputEV {
   type Props[EV[_], A]   = FormInputEV[EV, A]
   type ChangeCallback[A] = A => Callback
-  type Scope[EV[_], A]   = RenderScope[Props[EV, A], State, Unit]
 
   @Lenses
-  final case class State(curValue: String, prevValue: String)
-
-  def onTextChange[EV[_], A]($ : Scope[EV, A]): ReactEventFromInput => Callback =
-    (e: ReactEventFromInput) => {
-      // Capture the value outside setState, react reuses the events
-      val v = e.target.value
-      // First update the internal state, then call the outside listener
-      $.setStateL(State.curValue)(v) *>
-        // Next 2 might not be called if the InputFormat returns None
-        $.props.valSet(v) *>
-        $.props.onChangeC(v)
+  final case class State(curValue: String, prevValue: String, errors: Option[NonEmptyList[String]])
+
+  class Backend[EV[_], A]($ : BackendScope[Props[EV, A], State]) {
+
+    def validate(
+      props:     Props[EV, A],
+      value:     String
+    )(
+      onValid:   A => Callback = _ => Callback.empty,
+      onInvalid: NonEmptyList[String] => Callback = _ => Callback.empty
+    ): Callback = {
+      val validated     = props.validate.getValidated(value)
+      val validChangeCB = props.onValidChange(validated.isValid)
+      val onValidCB     = validated match {
+        case Valid(a)   => onValid(a)
+        case Invalid(e) => onInvalid(e)
+      }
+      validChangeCB >> onValidCB
     }
 
-  def onBlur[EV[_], A]($ : Scope[EV, A], c: ChangeCallback[String]): Callback =
-    c($.state.curValue)
+    def onTextChange(props: Props[EV, A]): ReactEventFromInput => Callback =
+      (e: ReactEventFromInput) => {
+        // Capture the value outside setState, react reuses the events
+        val v = e.target.value
+        // First update the internal state, then call the outside listener
+        $.setStateL(State.curValue)(v) >> $.setStateL(State.errors)(none) >>
+          validate(props, v)(a =>
+            /*props.valSet(a) >>*/ props.onChange(a) // TODO Maybe we don't call valSet here???
```

Comment:
If we don't call `valSet` the value isn't propagated with the view to the parent, which maybe is what we want.